### DISCUSSION TITLE
generate a version.text in obj folder when publish and make it to copy to publish folder

### DISCUSSION THREAD

**Discussion by John0King** [dotnet/msbuild#9160] [Q&A] [answered]
my project is a traditional asp.net mvc project.
I'm using "FileSystem" to publish the web site,  I successed generated the version.txt base a msbuild property 

```msbuild
<Target Name="_Generate_version_txt" BeforeTargets="ResolveFrameworkReferences" Condition="'$(MSBuildProjectName)' == 'MyProjectName' And '$(WebPublishMethod)' != ''  ">
  <PropertyGroup>
    <VersionFilePath>$([System.IO.Path]::Combine('$(IntermediateOutputPath)','version.txt'))</VersionFilePath>
  </PropertyGroup>

  <WriteLinesToFile File="$(VersionFilePath)" Lines="Version:$(VersionX)" Overwrite="true" />
  <!--<Copy SourceFiles="$(VersionFilePath)" DestinationFolder="$(OutputPath)"></Copy>-->
  <ItemGroup>
    <Content Include="@(VersionFilePath)">
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
      <CopyToPublishDirectory>Always</CopyToPublishDirectory>
      <DestinationRelativePath>Version.txt</DestinationRelativePath>
      <TargetPath>Version.txt</TargetPath>
    </Content>
  </ItemGroup>
</Target>
```

but I have trouble on make the `obj/version.txt` copy to publish folder , I tried using `copy`  task to copy it to publish folder directly, but then I find out that it will delete all the "additional" files when publish

**Comment by GangWang01** [comment]
`<Content Include="@(VersionFilePath)">` has a small mistake on the syntax referencing the property. It should be `$(VersionFilePath)` prefixing the property with `$` rather than `@`. 
The following cleaned target should work for your purpose.
```
<Target Name="_Generate_version_txt" BeforeTargets="ResolveFrameworkReferences" Condition="'$(MSBuildProjectName)' == 'MSBuildProjectName' And '$(WebPublishMethod)' != ''  ">
  <PropertyGroup>
    <VersionFilePath>$([System.IO.Path]::Combine('$(IntermediateOutputPath)','version.txt'))</VersionFilePath>
  </PropertyGroup>

  <WriteLinesToFile File="$(VersionFilePath)" Lines="Version:$(VersionX)" Overwrite="true" />
  <ItemGroup>
    <Content Include="$(VersionFilePath)">
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
      <TargetPath>Version.txt</TargetPath>
    </Content>
  </ItemGroup>
</Target>
```

**Reply by John0King** [reply]
it actually doesn't work ,  The asp.net project seems very special， some thoughts are :

a) because I generate it on `obj/Release/Version.txt` , so the publish step doesn't care (remove it from publish source),  
b) or  my target runs too later (ps. I use  `BeforeTargets="Build"`)

**Comment by John0King** [comment, marked as answer]
@GangWang01   Thank ,  and  I search the web today , then I found this micorosoft doc,  https://learn.microsoft.com/en-us/aspnet/web-forms/overview/deployment/visual-studio-web-deployment/deploying-extra-files   
it depends on a `<FilesForPackagingFromProject />` item , and `DestinationRelativePath`  metadata
